File: C++/Graphs/17_SlidingPuzzle.cpp
int slidingPuzzle(vector<vector<int>> &board)
{
    string Fstr = "123450";

    string curr = "";

    int idx = 0;
    for (int i = 0; i < 2; i++)
    {
        for (int j = 0; j < 3; j++)
        {
            curr += to_string(board[i][j]);
            if (board[i][j] == 0)
                idx = i * 3 + j;
        }
    }
    // cout<<curr<<" "<<idx<<endl;

    vector<vector<int>> dir{{-1, 0}, {1, 0}, {0, 1}, {0, -1}};

    if (curr == Fstr)
        return 0;
    // if(idx == 5 and curr!= Fstr)
    //     return -1;

    queue<pair<string, int>> q;
    q.push({curr, idx});

    map<string, bool> vis;
    vis[curr] = true;

    int ans = 0;

    while (!q.empty())
    {
        int qsize = q.size();
        while (qsize-- > 0)
        {
            auto p = q.front();
            string s = p.first;
            int ind = p.second;
            q.pop();

            if (s == Fstr)
            {
                return ans;
            }

            int x = ind / 3;
            int y = ind % 3;
            for (int d = 0; d < 4; d++)
            {
                int a = x + dir[d][0];
                int b = y + dir[d][1];
                if (a >= 0 and b >= 0 and a < 2 and b < 3)
                {
                    int newIndex = a * 3 + b;
                    string temp = s;
                    swap(temp[newIndex], temp[ind]);
                    if (!vis[temp])
                    {
                        q.push({temp, newIndex});
                        vis[temp] = true;
                    }
                }
            }
        }
        ans++;
    }

    return -1;
}

File: Java/Graphs/17_SlidingPuzzle.java
class Solution {
    public String swap(String s,int i,int j){
        char[] arr = s.toCharArray();
        
        char temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
        
        return new String(arr);
    }
    public int slidingPuzzle(int[][] board) {
        int n = board.length;
        int m = board[0].length;
        int dir[][] = {{-1,0},{0,-1},{0,1},{1,0}};
        
        String dest = "123450";
        StringBuilder sb = new StringBuilder();
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                sb.append(board[i][j]);
            }
        }
        String src = sb.toString();
        HashSet<String> vis = new HashSet<>();
        Queue<String> que = new LinkedList<>();
        que.add(src);
        vis.add(src);
        int moves = 0;
        
        while(que.size() != 0){
            int size = que.size();
            while(size-->0){
                String str = que.remove();
                
                if(str.equals(dest)) return moves;
                
                int idx = str.indexOf("0");
                int i = idx/m;
                int j = idx%m;
                
                for(int d=0;d<dir.length;d++){
                    int r = i + dir[d][0];
                    int c = j + dir[d][1];
                    
                    if(r>=0 && c>=0 && r<n && c<m){
                        String next = swap(str,idx,r*m+c);
                        if(!vis.contains(next)){
                            vis.add(next);
                            que.add(next);
                        }
                    }
                }
            }
            moves++;
        }
        return -1;
    }
}